### PR TITLE
fix(cache): daily-bar staleness backstop with single-flight refetch

### PR DESCRIPTION
**Pull request by Chen-zexi**
## Overview

Adds a daily-bar **staleness backstop** to the OHLCV cache and serializes the cold/stale full-fetch path so a burst of concurrent readers can't stampede the upstream provider.

The core bug: an envelope's `data_date` can be silently re-stamped to today by a delta refresh that fetched **no new bars** (provider hadn't published today's daily bar yet). `_is_stale_date` then reads "fresh" forever while the actual bars are days behind. This adds a watermark-level backstop that compares the newest bar's ET trading date against the newest daily bar that *should* exist right now — pre-market aware, since day D's daily bar only appears at the 09:30 ET open.

### Net change

| Category            | Files | +LOC | −LOC |
|---------------------|------:|-----:|-----:|
| Production          |     3 |  225 |   69 |
| Tests               |     3 |  391 |    4 |
| **Net (excl. tests)** |   3 | **+156** |     |

### By module

**`src/utils/market_hours.py`** (+37/−7)
- New `expected_latest_daily_date(now)` — the newest daily bar that should exist now (previous trading day during pre-market; today once the session opens).
- Extracted `_latest_trading_day(start, ok=None)` helper; `current_trading_date` refactored onto it.

**`src/server/services/cache/_ohlcv_envelope.py`** (+76/−7)
- `is_watermark_stale(..., symbol, is_index)` extended with a `1day` branch: date-level compare of the watermark's ET date vs `expected_latest_daily_date()`, gated to US-calendar symbols with a `_DAILY_STALE_REFETCH_COOLDOWN` (120s) floor, empty-bar skip, and corrupt-watermark handling.
- `_follows_us_daily_calendar(symbol, is_index)` — US-calendar classifier. Broadened so US dotted class-shares (BRK.B, BF.B) and an expanded US index allowlist (SOX, RUT, …) keep the backstop, while genuine foreign tickers (.HK/.SS/.L/.T) correctly skip it.
- The daily branch **fails closed without a symbol**: callers that can't classify the calendar anchor skip the backstop instead of being judged against the US calendar (guards the symbol-less intraday call sites if `1day` ever routes through them).

**`src/server/services/cache/daily_cache_service.py`** (+112/−55)
- Cold/stale full-fetch path serialized on a per-series `asyncio.Lock` (same idiom as `_delta_refresh`) with a **double-check** re-read after acquiring the lock: the leader fetches + fills, followers serve the fill instead of each firing their own blocking upstream fetch.
- Extracted `_cached_result` helper so the double-check and the hit path build identical cache-hit results.

### What this introduces

A daily-granularity staleness signal that survives a `data_date` re-stamp, plus single-flight protection against a thundering herd at the open — without changing behavior for fresh envelopes or non-US symbols.

## Test coverage

- `tests/unit/utils/test_market_hours.py` (new) — `current_trading_date` + `expected_latest_daily_date` across phases/weekends.
- `tests/unit/server/services/test_ohlcv_envelope_staleness.py` — daily backstop (stale/fresh/cooldown/corrupt/empty), US dotted class-share + index-allowlist retention, `_follows_us_daily_calendar` classification matrix (US bare/.US/dotted, foreign suffixes, index membership).
- `tests/unit/server/services/test_daily_cache_dedup.py` (new) — concurrent cold reads coalesce to one fetch; distinct symbols don't coalesce; leader cancellation doesn't poison the follower; second identical call serves cache.

## Review

Six review passes (2 Opus adversarial + 2 Codex per request + 1 Codex on the final lock rewrite + 1 independent re-derivation). Consensus surfaced two issues, both addressed:
- **Cache stampede (P1)** on the lockless cold/stale path → serialized with per-key lock + double-check.
- **Symbol-classification gaps (P2)** — US dotted class-shares and several US indexes were skipping the backstop → classifier broadened.
- A concurrency defect in the first (shared-task + `asyncio.shield`) dedup implementation — one waiter's cancellation poisoning all coalesced callers — was found by Codex and eliminated by switching to the lock approach (a cancelled waiter just releases the lock).
- The final re-derivation pass verified the backstop's three silent assumptions: US daily bars are stamped at ET-midnight-or-later across all three sources (FMP stamps `tzinfo=ET`, ginlix-data passes through Polygon's midnight-ET stamps, yfinance ≥1.2 returns exchange-localized indexes), the holiday table covers 2025–2027 with a rollover warning, and `SUPPORTED_INTERVALS` excludes `1day` so the symbol-less intraday call sites can't reach the daily branch. It also produced the fail-closed hardening for the no-symbol case.

## Known limitations

- The backstop is US-calendar only by design; non-US daily symbols continue to rely on `_is_stale_date` (their exchange-local midnight maps to the previous ET date and would read as permanently stale against the US calendar).
- `_refresh_locks` grows one entry per distinct `(symbol, from, to, is_index)`; unbounded in principle but bounded by the tracked-symbol universe in practice, same as the pre-existing `_delta_refresh` lock map.

## Test plan

- `uv run pytest tests/unit/server/services/ tests/unit/utils/ -q` — green (cache + utils suites).
- `uv run ruff check src/server/services/cache/ src/utils/market_hours.py` — clean.

## Summary by CodeRabbit

* **New Features**
  * Enhanced daily OHLCV freshness logic using market trading-day awareness (including pre-market, weekends/holidays, and a short post-fetch cooldown).
  * Improved handling of US-market vs non-US calendars to ensure freshness checks apply only where appropriate.

* **Bug Fixes**
  * Reduced unnecessary duplicate upstream fetches by coalescing concurrent requests for the same daily series.
  * More consistent treatment of empty or malformed daily cache envelopes to prevent incorrect staleness decisions.

* **Tests**
  * Added unit coverage for daily freshness edge cases and cache coalescing/concurrency behavior.

### CODE REVIEW

File: src/server/services/cache/_ohlcv_envelope.py [hunk 1]
```diff
@@ -133,16 +174,40 @@ def is_watermark_stale(
...
+            return False
+        if not envelope.get("bars"):
+            return False  # empty window — soft-TTL governs re-fetch timing
+        if time.time() - envelope.get("fetched_at", 0) < _DAILY_STALE_REFETCH_COOLDOWN:
+            # Just fetched — the provider simply hasn't published a newer bar
+            # yet; flagging stale again would re-fetch on every request.
+            return False
+        last_bar_date = watermark_to_date_str(envelope.get("watermark"))
+        if last_bar_date is None:
+            return True  # bars present but unusable watermark — corrupt
```
[the remark this answers is not in the document]
Comment (Chen-zexi) [reply src/server/services/cache/_ohlcv_envelope.py:211]: Rejected — the ordering is deliberate (now documented in e386b023). If the feed is persistently corrupt (last bar with no usable `time`), the refetched envelope is corrupt again with a fresh `fetched_at`. With the corrupt check first, every request flags stale, passes `_full_fetch`'s double-check (`_needs_refresh` stays True), and pays a blocking upstream fetch — the exact per-request stampede the cooldown exists to prevent. The current order bounds a corrupt feed to one refetch per 120s while serving the (present) bars in between; detection delay is capped at the cooldown and corruption is a terminal state, not a freshness signal, so the delay has no user-visible cost.

File: src/server/services/cache/daily_cache_service.py [hunk 2]
```diff
@@ -230,27 +231,31 @@ async def get_stock_daily(
...
-                if is_live and (_is_stale_date(envelope) or envelope.get("complete")):
-                    # Stale date or day-boundary transition → sync re-fetch
+            if _needs_refresh(
+                envelope, base_ttl, interval="1day", is_live=is_live,
+                symbol=normalized, is_index=is_index,
+            ):
+                if is_live and (
+                    _is_stale_date(envelope)
+                    or is_watermark_stale(envelope, "1day", symbol=normalized, is_index=is_index)
+                    or envelope.get("complete")
```
[the remark this answers is not in the document]
Comment (Chen-zexi) [reply src/server/services/cache/daily_cache_service.py:251]: Rejected. The 04:00 ET `data_date` flip triggers exactly one sync refetch per symbol per day: the refetch re-stamps `data_date`, and the new watermark backstop is pre-market-aware, so nothing flags again until 09:30. This behavior predates the PR (the `_is_stale_date`-gated sync path is pre-existing); the PR only narrowed pre-market refetching. That single fetch is also where overnight adjustments (splits/dividends rewriting historical bars) get picked up. A daily-aware predicate here would need a US-only fork — `expected_latest_daily_date()` is US-calendar, while non-US symbols rely on the 04:00 flip as their only daily refresh signal — added branching for one saved fetch per symbol per day.

File: tests/unit/server/services/test_daily_cache_dedup.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+
+@pytest.mark.asyncio
+async def test_concurrent_cold_reads_coalesce_to_one_fetch(fresh_service):
+    svc, provider, _cache, gate = fresh_service
+
+    tasks = [asyncio.create_task(svc.get_stock_daily("AAPL")) for _ in range(5)]
+    # Let the leader grab the lock and the rest queue behind it on the gate.
+    await asyncio.sleep(0.02)
+    gate.set()
```
[the remark this answers is not in the document]
Comment (Chen-zexi) [reply tests/unit/server/services/test_daily_cache_dedup.py:75]: Rejected — there is no race to fix. asyncio is cooperative and single-threaded: when the test awaits `sleep(0.02)`, the loop runs the just-created tasks to their first real suspension point (`gate.wait()` for the leader, the refresh lock for the follower) before the sleep's timer callback can resume the test. Every stub await resolves without external I/O, and CI load preempts the whole process, never individual tasks, so the interleaving is deterministic for any nonzero sleep — the assertions cannot flake from load.

**Comment by Chen-zexi** [comment]
Adopted the `_US_DOTTED_SUFFIXES` / `_SUFFIX_MAP` disjointness suggestion in e386b023: `TestFollowsUsDailyCalendar` now asserts `_US_DOTTED_SUFFIXES.isdisjoint(_SUFFIX_MAP)`, so a future single-letter region suffix (the map already carries `L`/`T`) fails a test instead of silently classifying foreign dotted symbols as US-calendar.
